### ISSUE TITLE
Redesign agent-optimize: parallel graph search over candidate merges + adaptive task subsetting + TDD-style micro-tests

### ISSUE THREAD

**Issue by OsherElhadad** [skillberry-ai/cap-evolve#434]
# Redesign agent-optimize: parallel graph search over candidate merges + adaptive task subsetting + TDD-style micro-tests

## Problem — evidence from run3 (`run_agentoptv4`, and the same shape in `run_agentoptv3`)

Two recent hosted `agent-optimize` runs on the tau2-bench airline capability produced a **null result**:

| run | baseline val | best val | test | candidates | pattern |
|---|---|---|---|---|---|
| `run_agentoptv3` | 0.667 | seed (no accept) | 0.500 | c1_verify, c2_toolguard(+solo), c3_finish, c5_guards_only, c6_onerule, c0_nulledit | 6 candidates, every one gated on **full val**, every one **rejected** |
| `run_agentoptv4` | 0.567 | seed (no accept) | 0.500 | cA_partial, cB_becabin, c0_null(x3) | same shape |

In both, `best_id == "seed"` — a diagnosed null, not a 0.0. The pattern across both:

1. **3–6 narrow, single-issue candidates per run**, each a small targeted edit (one verify step, one tool guard, one "finish" rule).
2. **Every candidate paid the full 30-task val split**, no `screen.py` tier-1/tier-2 subset triage visible in the events before the full-val gate.
3. **No merging** — no `integrate.py`/`funcmerge.py`/`merge_taskopt.py` call appears in either run dir. Candidates that individually moved different tasks were never combined into one artifact and re-gated.
4. **No per-task fan-out** (`taskeval.py`) and no `mechanisms.py` ledger entries — nothing suggests the driver ever narrowed to "these 3 tasks are the loss, spend the round's rollouts on those."
5. Both runs stopped at seed after ~5 iterations of "narrow single-issue candidate → full-val reject" and sealed.

**Important finding**: this is *not* a missing-capability gap. `skills/algorithms/agent-optimize/scripts/{screen,round,taskeval,mechanisms,integrate,funcmerge,merge_taskopt}.py` and the accompanying references (`algorithm.md`, `per-task-fanout.md`, `measured-lessons.md`, `edit-design-lessons.md`) already describe exactly the adaptive-subsetting, per-task-fanout, and merge-and-regate loop this issue asks for. **The hosted driver simply never exercised most of that surface.** Two runs is a small sample, but the shared shape (full-val-only, no merge, no subset, quick convergence to a null) across two different baselines is consistent with the loop defaulting to its cheapest-to-reach step (SKILL.md step 4, "Honest gate on FULL val") and skipping the harder-to-invoke accelerators (screen/round/per-task-fanout/merge), which live one reference-file deep and are framed as "optional."

This redesign therefore has two halves: (a) new capability where a real gap exists (TDD-style micro-tests below full-rollout cost), and (b) **restructuring the loop itself so the existing accelerators are the default path, not an opt-in a driver can skip under time pressure** — moving from "prose loop + optional helpers" to a loop whose *default* iteration graph is subset-first, fan-out, and merge-aware, with full-val-only-single-candidate as the fallback for when a round genuinely has one idea.

## Design goals (from the task brief)

1. The optimizer chooses its own strategy **and** a relevant subset of tasks per candidate (not always full val), with the subset choice visible/reported.
2. Multiple issues get multiple candidates concurrently, each evaluated only on tasks relevant to it; candidates get **merged**, and merges get evaluated — a graph search over candidate combinations, run in parallel, converging faster toward 90%+ val.
3. Small deterministic/fast test cases (TDD-style) reproduce each targeted failure **before** a full task rollout is spent on it, using a well-known benchmark/test-case schema rather than inventing one.
4. Draw on 2024-2026 prior art for automated agent/prompt optimization.

## Prior art reviewed

- **This repo's own `agent-optimize` references** (`algorithm.md`, `measured-lessons.md`, `per-task-fanout.md`) already cite and adapt:
  - **GEPA** (arXiv:2507.19457) — reflective prompt evolution over a Pareto frontier of per-instance scores; a cheap train-minibatch local gate (`sum(child) > sum(parent)`) before paying for the expensive step. `screen.py`'s tier ladder is explicitly this idea ported to val with variance awareness added.
  - **Arbor** (github.com/RUC-NLPIR/Arbor) — independent workers in separate worktrees, a cheap dev-signal screen, merge only what clears a bar. Its `merge_threshold` is log-only (never blocks) and has no repeated-trial/SE/significance machinery — `screen_decision`'s `Δ̄ + k·SE < 0` kill-only rule is the fix already applied.
  - The measured lessons on load-vs-noise, canary selection (draw from the *whole* suite, not the neighborhood of the target tasks — a canary set that only covers what you aimed at cannot catch what you hit by accident), and cross-occasion drift are directly relevant to point 2's "evaluate merges" step and should not be re-derived.
- **Harbor / Terminal-Bench** (github.com/harbor-framework/harbor, from the Terminal-Bench authors) — the schema to adopt for point 3. A Terminal-Bench-style task is a self-contained directory: `task.yaml`/`task.toml` (metadata + timeout), `instruction.md` (the scenario in natural language), an `environment/` (Dockerfile / minimal fixture state), a `solution/` reference fix, and a `tests/` dir whose `test.sh`/`run-tests.sh` asserts pass/fail deterministically and writes a reward (0/1 or partial) to a known path (e.g. `reward.txt`) — no LLM judge required for the pass/fail check itself, only for the task's own agent-under-test. Harbor generalizes this to "any benchmark" and runs it under `harbor run --dataset X --agent Y --model Z`. This is exactly the "small, deterministic, fast" shape point 3 wants: a micro-case is a miniature Terminal-Bench task, scoped to ONE observed failure, with a fixture instead of a full tau2 conversation and an assertion script instead of a full scorer.
- **Anthropic's public guidance on skill/agent eval design** — the pattern of building the smallest possible fixture that reproduces a failure and asserting on it programmatically (not via a second LLM judge) before spending a real end-to-end rollout, and of keeping eval cases as data (inputs + expected observable behavior) separate from the harness that runs them — both align with adopting the Harbor/Terminal-Bench shape rather than a bespoke one.
- **AlphaEvolve** (Google DeepMind, 2025) — evolutionary program search over an LLM-mutated population, using (a) an **evaluation cascade**: cheap/fast correctness checks filter the population before the expensive full evaluation runs, and (b) a **program/candidate database** organized as islands with periodic migration, trading exploration (diversity across islands) against exploitation (best-of-island refinement). Directly maps to: micro-tests as the cascade's cheap stage, and the candidate DAG (below) as the island-like structure with merge-as-migration.
- **DSPy optimizers** (MIPROv2, BootstrapFewShot, COPRO) — treat prompt/program optimization as a Bayesian-ish search over a discrete configuration space, using a train minibatch to score candidates cheaply before a full validation pass, and explicitly support **compositional programs** (multiple modules optimized jointly) — relevant to representing candidates as edits to named "slots" (files/functions) that can be recombined independently, the same unit `funcmerge.py` already merges on.
- **TextGrad** (arXiv:2406.07496) — backpropagates natural-language "gradients" (critiques) through a computation graph of LLM calls, so a system with multiple components gets per-component feedback rather than one global scalar. Relevant to diagnosing which *cluster* of failures maps to which *file/function*, sharpening what candidate should even be proposed for a given cluster — already partially present via `diagnose.py`'s clusters.
- **OPRO** (arXiv:2309.03409, "Large Language Models as Optimizers") — keeps an explicit history of (candidate, score) pairs in the meta-prompt so the optimizer LLM can see its own trajectory and avoid repeating discredited moves. Directly applicable to strengthening SKILL.md's existing "read `rejected.jsonl`, propose something structurally different" rule into a first-class optimization-trajectory object the freestyle agent consults every round, not just at proposal time.
- **APE — Automatic Prompt Engineer** (arXiv:2211.01910) — generate-many-then-select: propose a batch of candidate instructions, score all of them on a cheap subset, keep only the top-k for expensive evaluation. This is the "concurrent multi-candidate, cheap-subset-first" shape in goal 2, applied specifically to the *proposal* step rather than only the eval step.
- **Reflexion / Self-Refine** (2023) — verbal reinforcement: an agent reflects on *why* a trial failed in natural language and carries that reflection into the next attempt. Already the shape of `diagnose.py`'s per-cluster feedback; worth citing because it justifies keeping the reflective-dataset step even as the rest of the loop becomes more graph-like.

### 1. Data structure: a candidate DAG, not a flat list of siblings

Replace "propose N sibling candidates off the current best" with an explicit, persisted graph:

```
CandidateNode = {
  id: str,                        # unique tag, filename-safe
  parents: [str],                 # 1 for an edit, 2+ for a merge
  cluster_ids: [str],              # which diagnose() cluster(s) this targets
  edit_kind: "prompt" | "code" | "merge",
  micro_tests: [str],              # micro-test ids this was designed to pass (see #3)
  subset: {task_ids: [...], rationale: str, tier: int},  # see #2
  status: "proposed" | "micro_tested" | "screened" | "gated" | "accepted" | "rejected" | "superseded",
  val_mean: float | null,          # only set once FULL val has run (or null for subset-only)
  screen: {...} | null,            # screen.py output, if a screen ran
  gate: {...} | null,              # gate_check.py output, if the full gate ran
}
```

Persisted as `$R/graph.jsonl` (append-only, one event per state transition — same audit-log shape as `events.jsonl`, not a new invariant). `round.py`/`commit.py` gain a thin wrapper that also appends here; no change to what they already enforce (full-val-only gate, paired significance, no-regression veto, single test seal) — the DAG is a *view* over the existing accept/reject events plus the new merge/subset metadata, not a new authority.

This turns "merge and re-evaluate" into graph search: a round's frontier is the set of `accepted` or `screened-promote` nodes; the next round's candidates are either (a) a fresh edit off an accepted node, or (b) a **merge node** whose parents are 2+ nodes that individually screened well but target disjoint clusters.

### 2. Adaptive task subsetting, chosen by the optimizer and reported

Formalize what `screen.py` already computes (`broken_ids` → most-informative-remaining → seeded random holdout) as the **default per-candidate selection**, and require every candidate node to record *why* its subset looks the way it does:

```
subset.rationale examples:
  "targets cluster c3 (tasks 8,14,22,29 — timeout-before-tool-call); +6-task random holdout from parent's passing set as regression canary"
  "merge of node t7 (targets 8,14) + node t17 (targets 22,29); subset = union of both parents' targets + t7's holdout + t17's holdout"
```

This is not a new algorithm — it is (a) making `screen.py`'s tier-1/tier-2 ladder the thing every candidate goes through **by default**, not an optional accelerator a time-pressured driver skips, and (b) requiring the *rationale* to be written into the graph node so "which subset, and why" is inspectable without re-deriving it from `screens/*.json`. The freestyle agent still picks the subset (goal 1: "optimizer chooses ... with the subset choice reported/visible") — the change is that skipping this step becomes a visible gap in `graph.jsonl` (a node with `subset: null` going straight to a full-val gate), not an invisible one.

Full val remains the *only* thing that can accept (per the existing honesty invariant — `screen.py` has no accept path). What changes is the on-ramp: a candidate reaches the full-val gate only after (a) a micro-test (below) proves the mechanism fires, and (b) a subset screen fails to prove harm — collapsing most of the "narrow candidate → full 30-task val → reject" cycles seen in run3/run4 into a micro-test-fail or a screen-kill, at a fraction of the cost.

### 3. TDD-style micro-tests: adopt the Harbor/Terminal-Bench task schema, scoped to one failure

Before any rollout is spent on a candidate, generate a **micro-case** per targeted cluster — a miniature task in the same shape as a Terminal-Bench task, not a new invention:

```
$R/microcases/<cluster_id>/
  case.yaml        # {id, cluster_id, source_task_ids: [...], timeout_s, expects: "guard_fires"|"call_shape"|"tool_selected"}
  fixture/          # minimal conversation prefix / tool-call payload that reproduces the failure —
                     # extracted verbatim from the diagnosed rollout, never invented
  reproduce.py      # calls run_target (or a narrower unit — one tool function, one prompt turn)
                     # directly with the fixture, no full multi-turn tau2 episode
  assert.py         # deterministic pass/fail — exit 0/1, writes result.json, no LLM judge
```

This mirrors Terminal-Bench's `task.yaml` + `instruction`/environment + `tests/` split exactly (metadata, fixture state, deterministic assertion writing a reward), scoped down from "a whole task" to "the one call or turn where the diagnosed defect lives." It reuses the table already in `per-task-fanout.md` ("evidence that does not depend on load": *the guard fired on the observed call*, *the agent's next action changed after it fired*, *a direct call with the exact bad payload now succeeds/still refuses*) — this design just gives that evidence a standard on-disk shape and a place to accumulate (`microcases/`) instead of being reconstructed ad hoc per subagent report.

Cost: a micro-case runs in seconds (one call, no multi-turn episode, no full adapter rollout), versus a full task rollout. A candidate must pass its own micro-test(s) before it is allowed into a subset screen; a candidate whose micro-test fails is rejected on `--reject-basis infra`-equivalent (call it `micro_test_fail`) without spending any rollout at all. This is the "reproduce it small and fast before a real rollout" ask (goal 3), and it directly attacks the run3/run4 pattern of 5-6 rounds of "propose → full val → reject" with no cheaper filter in between.

Micro-cases are generated from `diagnose.py`'s cluster output (the failing rollout already names the tool call / turn where the defect shows), not invented from scratch, and are retained across rounds as a regression corpus — a candidate that breaks a *previously passing* micro-case is flagged before it ever reaches a screen.

### 4. Concurrent multi-issue candidates, then graph-search merge

Per round: for each cluster `diagnose.py` reports (up to an affordability-checked `N`), fan out one candidate proposal + micro-test **concurrently** (goal 2's "multiple candidates concurrently"), following the existing parallelism invariants in `algorithm.md` (unique tag per sibling, diagnosis/proposal fan freely, gate stays serial). Candidates that pass their micro-test and clear a subset screen become the round's **frontier nodes**.

Merging is then graph search over the frontier, using `integrate.py`/`funcmerge.py` exactly as documented in `per-task-fanout.md` — one branch at a time, re-measuring after each addition, never a single N-way merge — but driven as an explicit search rather than a manual one-off:

- Pairwise merge candidates whose target clusters are **disjoint** (files/functions don't overlap — `funcmerge.py`'s `dropped_additions` already detects unsafe overlap).
- Each merge node gets its own micro-tests = union of its parents' micro-tests (a merge must still pass every micro-test either parent passed alone — this is what catches the "clean merge is syntactic, composition is empirical" trap `per-task-fanout.md` already names).
- Each merge node gets its own subset = union of parents' target tasks + parents' holdouts + a fresh canary drawn from the *whole* suite (not the merge's own neighborhood — the exact canary-selection rule already in `per-task-fanout.md`, generalized from "per-task-fanout artifact" to "any merge node").
- Merges that clear their subset screen queue for the full-val gate like any other node; merges that fail feed back into the graph as `superseded`, with the reason (which parent's contribution regressed what) recorded via `mechanisms.py`, so the same collision isn't retried blind next round.

This is the graph search: nodes are candidates (edits or merges), edges are "parent of," the frontier advances by either targeted edit (new cluster) or merge (combine surviving edits), and full-val gates are paid only for nodes that already cleared the cheaper cascade (micro-test → subset screen). It converges faster than "N flat siblings, all full-val, never combined" because (a) more of the round's rollout budget buys distinct information (fewer full-val evals wasted on ideas a micro-test would have killed for free), and (b) fixes that are individually small (single-cluster) get to compound via merge instead of being evaluated, accepted or rejected, and forgotten one at a time.

### 5. What does NOT change (the honesty invariants stay exactly as documented)

- Full val is still the only thing that can **accept**; screens and micro-tests can only kill or gate cheaply.
- The paired significance gate (`Δ > k·SE`), the no-regression veto, and the single-use test seal are unchanged — the DAG and micro-tests are new information flowing into which candidates *reach* the gate, not a new way to pass it.
- `spend.py` affordability checks happen before every fan-out exactly as now, extended to price "N candidates + up to N·(N-1)/2 pairwise merges" rather than just N siblings, so a graph-search round can't blow a budget that looked affordable for a flat round.

## Why this addresses the run3/run4 null result specifically

The observed failure mode was: narrow single-cluster edits, each paying full val alone, never combined, exhausting the round budget on isolated small fixes that individually don't clear the significance bar. The redesign's cascade (micro-test → subset screen → merge → full val) means (a) a round can afford to explore more clusters because most of them get filtered before paying for a rollout, and (b) several individually-too-small fixes that would each fail their own full-val gate get a chance to combine into one artifact whose combined effect might clear the bar — which is structurally the thing that was completely absent from both observed runs.

## Phased implementation plan

This issue is the design. Implementation should land as separate, reviewable PRs against `skills/algorithms/agent-optimize/`, each with its own smaller tracking issue (linked below):

- **Phase 1 — Candidate DAG + graph.jsonl.** Add the append-only graph log and the thin wrapper around `commit.py`/`round.py` that writes to it. No behavior change to gating; purely observability, so it can land and be validated against run3/run4-style runs before anything else depends on it.
- **Phase 2 — Micro-test schema + generator.** Add the Harbor/Terminal-Bench-shaped `microcases/` layout, a generator that extracts a fixture from a diagnosed cluster's failing rollout, and the `reproduce.py`/`assert.py` contract. Wire "micro-test fail" as a new cheap reject-basis, before screen.py is touched.
- **Phase 3 — Default subset-first on-ramp.** Make `screen.py`'s tier ladder the default path SKILL.md's step 2→3 walks through (rather than an optional accelerator), and require `subset.rationale` on every candidate node from Phase 1.
- **Phase 4 — Merge-as-graph-search.** Extend `round.py`/`integrate.py` orchestration to auto-propose pairwise merges among a round's disjoint-cluster survivors, gated through the same cascade, with canary selection and `mechanisms.py` bookkeeping as already documented in `per-task-fanout.md`.

Each phase is independently valuable and independently testable against the existing `run_agentoptv3`/`run_agentoptv4` run dirs as regression fixtures (re-run the same baseline+diagnose state through the new on-ramp and check the round produces more graph-search structure and fewer flat full-val rejects for the same rollout budget).
